source/differentiation/ad/ad_drivers.cc

```diff
@@ -0,0 +1,2260 @@
+      // https://gitlab.com/adol-c/adol-c/blob/master/ADOL-C/src/tape_handling.cpp
+      // An alternative solution would be to manually access the tape data;
+      // see the removeTape() function in
+      // https://gitlab.com/adol-c/adol-c/blob/master/ADOL-C/src/tape_handling.cpp
+      // with the consideration of the #defines in
+      // https://gitlab.com/adol-c/adol-c/blob/master/ADOL-C/src/taping_p.h
+      // as to how this would be performed.
+      // Doing things "manually" (the second way) without creating the
+      // additional data object would be a lot more work...
+      //
+      // Both of the above solutions would be possible IF ADOL-C exposed
+      // this data object or a method to access it to the outside world.
+      // But they don't :-(
+      // Instead, what we'll have to do is get the statistics for this tape
+      // and make our own determination as to whether or not this tape exists.
+      // This effectively executes the first solution, with even more
+      // overhead! If the tape is in existence, we can assume that it should a
+      // non-zero number of dependent and independent variables. Those are
+      // stored in the zeroth and first entries of the statistics vector.
+      //
+      // But, oh wait... Surprise! This will trigger an error if the tape
+      // doesn't exist at all! So lets first check their tape info cache to
+      // see if the tape REALLY exists (i.e. has been touched, even if nothing
+      // has been written to it) before trying to access it. It'll only take
+      // an O(n) search, but at this point who really cares about efficiency?
+      //
+      // It has been suggested in
+      // https://gitlab.com/adol-c/adol-c/issues/11
+      // that a simply try-catch block around ::tapestats is the solution that
+      // we want here. Unfortunately this results in unwanted pollution of
+      // the terminal, of the form
+      // ADOL-C error: reading integer tape number 4!
+      //               >>> File or directory not found! <<<
+      // , every time a query is made about a non-existant tape.
+      // So either way we have to guard that check with something more
+      // conervative so that we don't output useless messages for our users.
+      const std::vector<typename Types<ADNumberType>::tape_index>
+                 registered_tape_indices = get_registered_tape_indices();
+      const auto it = std::find(registered_tape_indices.begin(),
+                                registered_tape_indices.end(),
+                                tape_index);
+      if (it == registered_tape_indices.end())
+        return false;
+
+      // See https://gitlab.com/adol-c/adol-c/issues/11#note_108341333
+      try
+        {
+          std::vector<std::size_t> counts(STAT_SIZE);
+          ::tapestats(tape_index, counts.data());
+          return true;
+        }
+      catch (const ::FatalError &exc)
+        {
+          return false;
+        }
+    }
```

Comment:
This is the fix for #7372 (i.e. reverting this function definition to its state before the last change.)